torch/_inductor/fx_passes/decompose_mm.py

```diff
@@ -0,0 +1,569 @@
+    non_zero_dim_reduce_scatter_pattern_single_user = reduce_scatter_template(
+        CallFunction(
+            aten.cat.default,
+            ListOf(
+                CallFunction(
+                    operator.getitem,
+                    CallFunction(
+                        aten.split.Tensor,
+                        KeywordArg("input"),
+                        Ignored(),
+                        KeywordArg("scatter_dim"),
+                        _users=MULTIPLE,
+                    ),
+                    Ignored(),
+                )
+            ),
+        ),
+        users=1,
+    )
+
+    zero_dim_reduce_scatter_pattern_single_user = reduce_scatter_template(
+        KeywordArg("input"), users=1
+    )
+
+    reduce_scatters = []
+    for node in reversed(g.nodes):
+        if node.target == c10d.wait_tensor.default:
+            if _config_dist.split_matmul_reducescatter_split_cat and (
+                match := non_zero_dim_reduce_scatter_pattern_single_user.match(node)
+            ):
+                assert isinstance(match, Match)
+                reduce_scatters.append(
+                    _ReduceScatterMatch(
+                        match=match,
+                        input_node=match.kwargs["input"],
+                        reduce_scatter_node=match.nodes[-2],
+                        wait_tensor_node=node,
+                        reduce_op=match.kwargs["reduce_op"],
+                        scatter_dim=match.kwargs["scatter_dim"],
+                        group_size=match.kwargs["group_size"],
+                        group_name=match.kwargs["group_name"],
+                    )
+                )
+                continue
+
+            if match := zero_dim_reduce_scatter_pattern_single_user.match(node):
+                assert isinstance(match, Match)
+                reduce_scatters.append(
+                    _ReduceScatterMatch(
+                        match=match,
+                        input_node=match.kwargs["input"],
+                        reduce_scatter_node=match.nodes[0],
+                        wait_tensor_node=node,
+                        reduce_op=match.kwargs["reduce_op"],
+                        scatter_dim=0,
+                        group_size=match.kwargs["group_size"],
+                        group_name=match.kwargs["group_name"],
+                    )
+                )
+                continue
+
+    for reduce_scatter in reversed(reduce_scatters):
+        _process_matmul_reduce_scatter(
+            g, reduce_scatter, min_size_after_split, num_chunks
+        )
+
+    trace_structured(
+        "artifact",
+        metadata_fn=lambda: {
+            "name": "pass_split_mm_rs_graph_after",
+            "encoding": "string",
+        },
+        payload_fn=lambda: gm.print_readable(
+            print_output=False, include_stride=True, include_device=True
+        ),
+    )
+
+
+@dataclasses.dataclass
+class _PostMMOp:
```

Comment:
cc @shunting314 , this is very similar to the work you have with chunking https://github.com/pytorch/pytorch/pull/136702/files#diff-b88e963821b3440e222c164d9b8a68143dbc92e33162084eae78c9cf46599713R86-R98 . would be great to consolidate eventually.